tensorboard/data/server/tf_record.rs

```diff
@@ -0,0 +1,327 @@
+    /// buf.extend(b"\x18\x00\x00\x00\x00\x00\x00\x00"); // length: 24 bytes
+    /// buf.extend(b"\xa3\x7f\x4b\x22"); // length checksum (0x224b7fa3)
+    /// let contents = b"\x09\x00\x00\x80\x38\x99\xd6\xd7\x41\x1a\x0dbrain.Event:2";
+    /// buf.extend(&contents[..5]); // file truncated mid-write
+    ///
+    /// let mut state = TfRecordState::new();
+    ///
+    /// // First attempt: read what we can, then encounter truncation.
+    /// assert!(matches!(
+    ///     state.read_record(&mut Cursor::new(buf)),
+    ///     Err(ReadRecordError::Truncated)
+    /// ));
+    ///
+    /// let mut buf: Vec<u8> = Vec::new();
+    /// buf.extend(&contents[5..]); // rest of the payload
+    /// buf.extend(b"\x12\x4b\x36\xab"); // data checksum (0xab364b12)
+    ///
+    /// // Second read: read the rest of the record.
+    /// let record = state.read_record(&mut Cursor::new(buf)).unwrap();
+    /// assert_eq!(record.data, contents);
+    /// assert_eq!(record.checksum(), Ok(()));
+    /// ```
+    pub fn read_record<R: Read>(&mut self, reader: &mut R) -> Result<TfRecord, ReadRecordError> {
+        if self.header.len() < HEADER_LENGTH {
+            read_remaining(reader, &mut self.header)?;
+
+            let (length_buf, length_crc_buf) = self.header.split_at(LENGTH_CRC_OFFSET);
+            let length_crc = MaskedCrc(LittleEndian::read_u32(length_crc_buf));
+            let actual_crc = MaskedCrc::compute(length_buf);
+            if length_crc != actual_crc {
+                return Err(ReadRecordError::BadLengthCrc(ChecksumError {
+                    got: actual_crc,
+                    want: length_crc,
+                }));
+            }
+
+            let length = LittleEndian::read_u64(length_buf);
+            let data_plus_footer_length_u64 = length + (FOOTER_LENGTH as u64);
+            let data_plus_footer_length = data_plus_footer_length_u64 as usize;
+            if data_plus_footer_length as u64 != data_plus_footer_length_u64 {
+                return Err(ReadRecordError::TooLarge(length));
+            }
+            self.data_plus_footer.reserve_exact(data_plus_footer_length);
+        }
+
+        if self.data_plus_footer.len() < self.data_plus_footer.capacity() {
+            read_remaining(reader, &mut self.data_plus_footer)?;
+        }
+
+        let data_length = self.data_plus_footer.len() - FOOTER_LENGTH;
+        let data_crc_buf = self.data_plus_footer.split_off(data_length);
+        let data = std::mem::take(&mut self.data_plus_footer);
+        let data_crc = MaskedCrc(LittleEndian::read_u32(&data_crc_buf));
+        self.header.clear(); // reset; caller may use this again
+        Ok(TfRecord { data, data_crc })
+    }
+}
+
+/// Fills `buf`'s remaining capacity from `reader`, or fails with `Truncated` if the reader is dry.
+fn read_remaining<R: Read>(reader: &mut R, buf: &mut Vec<u8>) -> Result<(), ReadRecordError> {
+    let want = buf.capacity() - buf.len();
+    reader.take(want as u64).read_to_end(buf)?;
```

Comment:
Here’s the `?` operator again. This time, `.read_to_end(buf)` is
a `Result<_, io::Error>`, but note that the enclosing function returns a
`Result<_, ReadRecordError>`—the types don’t line up. So the `?` on this
line is equivalent to:

```rust
match reader.take(want as u64).read_to_end(buf) {
    Err(io_error) => Err(ReadRecordError::from(io_error)),
    Ok(_) => (),
}
```

This is what the `impl From<io::Error> for ReadRecordError` above is
for: it describes how to convert an `io::Error` to our own error type,
so that we can use the `?` operator here.

Reply:
Interesting. I tend to think explicit statements (in this case, `::from` call) to be a good thing but I guess I will grow to like this.